cognify/optimizer/trace/checkpoint.py

```diff
@@ -0,0 +1,409 @@
+        
+        logs = self.layer_stats[layer_instance].load_existing_logs()
+        # update global best score and lowest cost
+        local_best_score = self.layer_stats[layer_instance].best_score
+        local_lowest_cost = self.layer_stats[layer_instance].lowest_cost
+        fast_exec_time = self.layer_stats[layer_instance].fastest_exec_time
+        self._update_glob_best_score(local_best_score)
+        self._update_glob_lowest_cost(local_lowest_cost)
+        self._update_glob_fastest_exec_time(fast_exec_time)
+        self.total_opt_cost += self.layer_stats[layer_instance].opt_cost
+        return logs
+    
+    def get_global_summary(self, verbose: bool):
+        candidates = []
+        for layer_instance, layer_stat in self.layer_stats.items():
+            if layer_stat.is_leaf:
+                candidates.extend(layer_stat.filter_by_constraint())
+        pareto_frontier = get_pareto_front(candidates)
+        if verbose:
+            print(f"================ Optimization Results =================")
+            print(f"Optimized for: {str(CommonStats.objectives)}")
+
+            if len(pareto_frontier) == 0:
+                print("Based on current optimization parameters, the best solution is the original workflow.")
+                print("We recommend increasing the number of trials or relaxing constraints.")
+            else:
+                print(f"Number of Optimized Workflows Generated: {len(pareto_frontier)}")
+
+            for i, trial_log in enumerate(pareto_frontier):
+                print("--------------------------------------------------------")
+                print("Optimization_{}".format(i + 1))
+                # logger.info("  Params: {}".format(trial_log.params))
+                score, price, exec_time = trial_log.result.reduced_score, trial_log.result.reduced_price, trial_log.result.reduced_exec_time
+                if CommonStats.base_quality is not None:
+                    print("  Quality improvement: {:.0f}%".format(
+                        _report_quality_impv(score, CommonStats.base_quality)
+                    ))
+                if CommonStats.base_price is not None:
+                    print("  Cost: {:.2f}x original".format(
+                        _report_cost_reduction(price, CommonStats.base_price)
+                    ))
+                if CommonStats.base_exec_time is not None:
+                    print("  Execution time: {:.2f}x original".format(
+                        _report_latency_reduction(exec_time, CommonStats.base_exec_time)
+                    ))
+                print("  Quality: {:.2f}, Cost per 1K invocation: ${:.2f}, Execution time: {:.2f}s".format(
+                    score, price * 1000, exec_time
+                ))
+                    # print("  Applied at: {}".format(trial_log.id))
+                    # logger.info("  config saved at: {}".format(log_path))
+
+                # logger.info("Opt Cost: {}".format(self.opt_cost))
```

Comment:
Remove? In general, we shouldn't have commented code unless it serves some very specific purpose